### PR TITLE
Windows: MSVC: Allocate wave_slice_dq array using mem_alloc_dq()

### LINKED ISSUE #6466: JWST wheel fails to build >1.3.1 due to cube_build
Hi all!

I've been trying to track down a build error in JWST, and it appears to have started post 1.3.1 (AKA 1.3.1 is the last version I can install successfully. Appears to be in [cube_dq_utils.c](https://github.com/spacetelescope/jwst/blob/master/jwst/cube_build/src/cube_dq_utils.c), but the error persists going back to 1.3.2, but in cube_match_sky.c. Tracebacks are provided below:

## Environment:
Python 3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit (AMD64)] on win32
OS: Windows 11 21H2 22000.258

## Installation on main:
```
  building 'jwst.cube_build.cube_match_sky_pointcloud' extension
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DNUMPY=1 -IC:\Users\Duy\AppData\Local\Temp\pip-build-env-kbtb12tx\overlay\Lib\site-packages\numpy\core\include -IE:\STScI\gitRepos\astrogrism\envmain\include -IC:\Program Files\Python\Python309\include -IC:\Program Files\Python\Python309\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\ATLMFC\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /Tcjwst/cube_build/src/cube_dq_utils.c /Fobuild\temp.win-amd64-3.9\Release\jwst/cube_build/src/cube_dq_utils.obj
  cube_dq_utils.c
  jwst/cube_build/src/cube_dq_utils.c(109): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_dq_utils.c(110): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_dq_utils.c(139): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_dq_utils.c(140): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_dq_utils.c(522): error C2057: expected constant expression
  jwst/cube_build/src/cube_dq_utils.c(522): error C2466: cannot allocate an array of constant size 0
  jwst/cube_build/src/cube_dq_utils.c(522): error C2133: 'wave_slice_dq': unknown size
  jwst/cube_build/src/cube_dq_utils.c(610): error C2057: expected constant expression
  jwst/cube_build/src/cube_dq_utils.c(610): error C2466: cannot allocate an array of constant size 0
  jwst/cube_build/src/cube_dq_utils.c(610): error C2133: 'wave_slice_dq': unknown size
```
## Installation error on 1.3.2:
```
  building 'jwst.cube_build.cube_match_sky' extension
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -DNUMPY=1 -IC:\Users\Duy\AppData\Local\Temp\pip-build-env-5gl9lryc\overlay\Lib\site-packages\numpy\core\include -IE:\STScI\gitRepos\astrogrism\envmain\include -IC:\Program Files\Python\Python309\include -IC:\Program Files\Python\Python309\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\ATLMFC\include -IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /Tcjwst/cube_build/src/cube_match_sky.c /Fobuild\temp.win-amd64-3.9\Release\jwst/cube_build/src/cube_match_sky.obj
  cube_match_sky.c
  C:\Users\Duy\AppData\Local\Temp\pip-build-env-5gl9lryc\overlay\Lib\site-packages\numpy\core\include\numpy\npy_1_7_deprecated_api.h(14) : Warning Msg: Using deprecated NumPy API, disable it with #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
  jwst/cube_build/src/cube_match_sky.c(199): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_match_sky.c(200): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_match_sky.c(206): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_match_sky.c(207): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
  jwst/cube_build/src/cube_match_sky.c(612): error C2057: expected constant expression
  jwst/cube_build/src/cube_match_sky.c(612): error C2466: cannot allocate an array of constant size 0
  jwst/cube_build/src/cube_match_sky.c(612): error C2133: 'wave_slice_dq': unknown size
  jwst/cube_build/src/cube_match_sky.c(700): error C2057: expected constant expression
  jwst/cube_build/src/cube_match_sky.c(700): error C2466: cannot allocate an array of constant size 0
  jwst/cube_build/src/cube_match_sky.c(700): error C2133: 'wave_slice_dq': unknown size
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
``

### PR DESCRIPTION
**Pull request by jhunkeler**
Resolves _N/A_ 

Closes #6466, closes #6905

Related: #5578, https://github.com/spacetelescope/crds/issues/790#issuecomment-1261891494

This replaces two variable length array declarations that are preventing Microsoft CL from succeeding. (Refer to #6466 above.)

I tested locally with:
- Windows 10 Pro
- MSVC 2022
- Mambaforge

```
$ pytest -rsv jwst/cube_build/
============================= test session starts ==============================
platform linux -- Python 3.10.9, pytest-7.2.2, pluggy-1.0.0
crds_context: jwst_1064.pmap
rootdir: /home/jhunk/Downloads/omg/jwst, configfile: setup.cfg
plugins: asdf-2.14.3, requests-mock-1.10.0, openfiles-0.5.0, doctestplus-0.12.1, cov-4.0.0, ci-watson-0.6.1, jwst-1.9.5.dev37+g73135ae5b.d20230313
collected 16 items                                                             

jwst/cube_build/tests/test_configuration.py .....                        [ 31%]
jwst/cube_build/tests/test_cube_build_step.py .                          [ 37%]
jwst/cube_build/tests/test_miri_cubepars.py ...                          [ 56%]
jwst/cube_build/tests/test_nirspec_cubepars.py .                         [ 62%]
jwst/cube_build/tests/test_wcs.py ......                                 [100%]

============================== 16 passed in 3.48s ==============================
```

PS - I also noticed drizzle-1.13.7 fails to build on Windows, however at the time of this writing `drizzle/master` and `drizzle/1.13.7` are the same... yet `pip install git+https://github.com/spacetelescope/drizzle` actually succeeds. Very strange. Maybe the archive is missing something?

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Review by pllim — APPROVED** [review]
CI is also failing on `master`, so the failures here probably unrelated.

**Review by mcara — APPROVED** [review]

**Comment by jemorrison** [comment]
I have a similar fix to some other cube_build updates on a branch I am working on  and  I have tested it quite a lot. This fix cleans up the remaining memory leaks.

**Comment by mcara** [comment]
I do not think the old code could have resulted in a memory leak but allocating large arrays on stack is not a good idea in general and allocating variable size arrays is not supported by all compilers (not a standard C feature depending on the standard version). For MS, in particular see https://devblogs.microsoft.com/cppblog/c11-and-c17-standard-support-arriving-in-msvc/#variable-length-arrays

**Comment by jemorrison** [comment]
Running on a Mac M1 computer - the memory is smaller now than before the change to the dq arrays.

**Comment by pllim** [comment]
Just merge this already! 😸

**Comment by jemorrison** [comment]
I have a pr with exactly the same updates + other stuff. So if you merge this one I will rebase my code to grab you merge. It would helpful to merge this one soon because I want to submit my pr today and I can get it clean up with including this pr before I do that.

**Comment by pllim** [comment]
I don't have the power. Otherwise, I would have done it already. 😸

File: CHANGES.rst [hunk 1]
```diff
@@ -17,6 +17,11 @@ calwebb_detector1
 - Added the call to the undersampling_correction step to the ``calwebb_detector1``
   pipeline. [#7501]
 
+cube_build
+----------
+
+-  Windows: MSVC: Allocate ``wave_slice_dq`` array using ``mem_alloc_dq()`` [#7491]
```
Comment (hbushouse) [inline CHANGES.rst:23]: Just curious, what does MSVC stand for?

**Review by hbushouse — APPROVED** [review]

File: CHANGES.rst (hunk 1 above)
Comment (mcara) [reply CHANGES.rst:28]: https://en.m.wikipedia.org/wiki/Microsoft_Visual_C

Comment (mcara) [reply CHANGES.rst:28]: Comment could be improved

**Comment by pllim** [comment]
Thanks!

**Comment by pllim** [comment]
@zacharyburnett , are you sure jwst won't work on Windows? Though hard to tell without actually running CI on it.

xref https://github.com/spacetelescope/hstcal/issues/517

**Comment by zacharyburnett** [comment]
> @zacharyburnett , are you sure jwst won't work on Windows? Though hard to tell without actually running CI on it.
> 
> xref [spacetelescope/hstcal#517](https://github.com/spacetelescope/hstcal/issues/517)

Oh nice, I didn't know about this! Shows what I know haha 😅 

I can add some CI tests to the weekly cron to see how stable Windows support is

**Pull request merged**